### PR TITLE
Add modern logs for Serverless Framework v3

### PR DESCRIPTION
**Pull request by pgrzesik**
Configured in a way that would allow support both for v2 and v3 versions of the Framework

### CODE REVIEW

File: lib/clean.js [hunk 1]
```diff
@@ -38,8 +38,15 @@ function cleanup() {
 function cleanupCache() {
   const cacheLocation = getUserCachePath(this.options);
   if (fse.existsSync(cacheLocation)) {
+    let cleanupProgress;
     if (this.serverless) {
-      this.serverless.cli.log(`Removing static caches at: ${cacheLocation}`);
+      if (this.progress && this.log) {
```
Comment (medikoo) [inline lib/clean.js:43]: It's fine to check just for `this.log` (also later, when we may want to unconditionally switch to modern logs, it'll make easier to pick all those conditionals by searching `if (this.log`)

File: lib/clean.js [hunk 2]
```diff
@@ -50,10 +57,19 @@ function cleanupCache() {
       .forEach((file) => {
         promises.push(fse.removeAsync(file));
       });
-    return BbPromise.all(promises);
+    return BbPromise.all(promises)
+      .then(() => cleanupProgress && cleanupProgress.remove())
```
Comment (medikoo) [inline lib/clean.js:61]: Let's use `.finally`

File: lib/docker.js [hunk 3]
```diff
@@ -171,7 +189,7 @@ async function getDockerUid(bindPath) {
     '/bin/sh',
   ];
   const ps = await dockerCommand(options);
-  return ps.stdout.trim();
+  return ps.stdoutBuffer.trim();
```
Comment (medikoo) [inline lib/docker.js:192]: I take it's a fix to previous PR, not related to modern logs?

File: lib/inject.js [hunk 4]
```diff
@@ -92,10 +92,18 @@ function injectAllRequirements(funcArtifact) {
...
+    injectProgress.update('Injecting required Python packages to package');
+    this.log.info('Injecting required Python packages to package');
+  } else {
+    this.serverless.cli.log('Injecting required Python packages to package...');
+  }
 
+  let returnPromise;
   if (this.serverless.service.package.individually) {
-    return BbPromise.resolve(this.targetFuncs)
+    returnPromise = BbPromise.resolve(this.targetFuncs)
```
Comment (medikoo) [inline lib/inject.js:106]: Why not simply convert it to async function, and use `finally`?

File: lib/layer.js [hunk 5]
```diff
@@ -53,9 +53,23 @@ function layerRequirements() {
...
+    this.log.info('Packaging Python Requirements Lambda Layer');
+  } else {
+    this.serverless.cli.log('Packaging Python Requirements Lambda Layer...');
+  }
 
-  return BbPromise.bind(this).then(zipRequirements).then(createLayers);
+  return BbPromise.bind(this)
+    .then(zipRequirements)
+    .then(createLayers)
+    .then(() => layerProgress && layerProgress.remove())
```
Comment (medikoo) [inline lib/layer.js:68]: `.finally`

File: lib/zip.js [hunk 6]
```diff
@@ -104,21 +122,46 @@ function packRequirements() {
...
+            );
+          }
           f.package.patterns.push(`${f.module}/.requirements.zip`);
-          return addTree(
-            new JSZip(),
-            `.serverless/${f.module}/requirements`
-          ).then((zip) => writeZip(zip, `${f.module}/.requirements.zip`));
+          return addTree(new JSZip(), `.serverless/${f.module}/requirements`)
+            .then((zip) => writeZip(zip, `${f.module}/.requirements.zip`))
+            .then(() => packProgress && packProgress.remove())
```
Comment (medikoo) [inline lib/zip.js:142]: `.finally`

File: lib/inject.js (hunk 4 above)
Comment (pgrzesik) [reply lib/inject.js]: great suggestion

File: lib/docker.js (hunk 3 above)
Comment (pgrzesik) [reply lib/docker.js:192]: yes, I didn't spot it previously

File: lib/inject.js [hunk 7]
```diff
@@ -86,57 +86,70 @@ function moveModuleUp(source, target, module) {
...
+  }
 
-  if (this.serverless.service.package.individually) {
-    return BbPromise.resolve(this.targetFuncs)
-      .filter((func) =>
-        (func.runtime || this.serverless.service.provider.runtime).match(
-          /^python.*/
+  try {
+    if (this.serverless.service.package.individually) {
+      await BbPromise.resolve(this.targetFuncs)
```
Comment (medikoo) [inline lib/inject.js:106]: It should `return await` to work in same way (but not sure, maybe it won't make difference)

File: lib/inject.js [hunk 8]
```diff
@@ -86,57 +86,70 @@ function moveModuleUp(source, target, module) {
...
+          return this.options.zip
+            ? func
+            : injectRequirements(
+                path.join('.serverless', func.module, 'requirements'),
+                func.package.artifact,
+                this.options
+              );
+        });
+    } else if (!this.options.zip) {
+      await injectRequirements(
```
Comment (medikoo) [inline lib/inject.js:145]: Same here

File: lib/zip.js [hunk 9]
```diff
@@ -104,21 +122,42 @@ function packRequirements() {
...
+      }
       this.serverless.service.package.patterns.push('.requirements.zip');
-      return addTree(new JSZip(), '.serverless/requirements').then((zip) =>
-        writeZip(zip, path.join(this.servicePath, '.requirements.zip'))
-      );
+      return addTree(new JSZip(), '.serverless/requirements')
+        .then((zip) =>
+          writeZip(zip, path.join(this.servicePath, '.requirements.zip'))
+        )
+        .then(() => packProgress && packProgress.remove())
```
Comment (medikoo) [inline lib/zip.js:156]: `finally`

Comment (pgrzesik) [reply lib/zip.js]: good catch

File: lib/inject.js (hunk 7 above)
Comment (pgrzesik) [reply lib/inject.js:106]: I've skipped that return because the result is not actually used anywhere

**Review by medikoo — APPROVED** [review]
Great work 👍

**Pull request merged**